### PR TITLE
Fix macos bundle

### PR DESCRIPTION
**Pull request by hmueller01**
Fixed a few things to make build work on macOS.

### CODE REVIEW

**Comment by hmueller01** [comment]
Using Qt6 complies and creates the app as expected.

But Qt5 does not compile ATM. I get 
```
> make
[  0%] Built target nomacsCore-3.19.0_autogen_timestamp_deps
[  0%] Automatic MOC for target nomacsCore-3.19.0
[  0%] Built target nomacsCore-3.19.0_autogen
[  1%] Generating qrc_nomacs.cpp
[  2%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/nomacsCore-3.19.0_autogen/mocs_compilation.cpp.o
[  3%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkBaseWidgets.cpp.o
[  3%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkBatch.cpp.o
[  4%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkCentralWidget.cpp.o
[  5%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkConnection.cpp.o
[  6%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkControlWidget.cpp.o
[  6%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkDialog.cpp.o
[  7%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkDockWidgets.cpp.o
[  8%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkLogWidget.cpp.o
[  9%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkManipulatorWidgets.cpp.o
[ 10%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkMenu.cpp.o
[ 10%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkMetaDataWidgets.cpp.o
[ 11%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkNetwork.cpp.o
[ 12%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkNoMacs.cpp.o
[ 13%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkPong.cpp.o
[ 13%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkPreferenceWidgets.cpp.o
[ 14%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkQuickAccess.cpp.o
[ 15%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkSettingsWidget.cpp.o
[ 16%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkThumbsWidgets.cpp.o
[ 17%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkToolbars.cpp.o
[ 17%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkViewPort.cpp.o
[ 18%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkWidgets.cpp.o
[ 19%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkActionManager.cpp.o
[ 20%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkBaseViewPort.cpp.o
[ 20%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkBasicLoader.cpp.o
/Users/hmueller/src/opensource/nomacs.hmueller01/ImageLounge/src/DkCore/DkBasicLoader.cpp:76:10: fatal error: 'libraw/libraw.h' file not found
#include <libraw/libraw.h>
         ^~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkBasicLoader.cpp.o] Error 1
make[1]: *** [CMakeFiles/nomacsCore-3.19.0.dir/all] Error 2
make: *** [all] Error 2
```
This is strange as it is available
```
> ll /usr/local/include/libraw/libraw.h
-rw-r--r--  1 hmueller  admin  18208 Dec 19  2023 /usr/local/include/libraw/libraw.h
```

File: README.md [hunk 1]
```diff
@@ -132,22 +132,33 @@ cd build
...
-Qt5_DIR=/opt/homebrew/opt/qt5/ cmake -DQT_QMAKE_EXECUTABLE=/opt/homebrew/opt/qt5/bin/qmake ../ImageLounge/.
+$ Qt5_DIR=/opt/homebrew/opt/qt5/ cmake -DQT_QMAKE_EXECUTABLE=/opt/homebrew/opt/qt5/bin/qmake --install-prefix /Applications ../ImageLounge/.
 ```
 
 Run make:
 
 ```console
-make
+$ make
+$ make bundle
```
Comment (scrubbbbs) [inline README.md:152]: - I had a working bundle without `make bundle`. I think this should be part of the install step

**Comment by scrubbbbs** [comment]
@hmueller01 Thanks for contributing. The build for other platforms seems to be OK and you've moved the needle for OS X so I'm ready to merge as-is.  I leave it to you if you want to work on the remaining issues (if that's possible) in this PR.

After compiling your branch, `make` works, but `make bundle` fails for me. Some of the shared libs are not incorporated. I'm not sure what else. So we still may have a broken bundle.

I am compiling using a virtual machine I setup last year,  OS X 11.7.8 / homebrew v4.3.12 so that might be the reason. At that time, I had to write [my own deployment tool](https://github.com/scrubbbbs/cbird/blob/main/mac/mac-pkg.sh) for os x, probably because macdeployqt wasn't working at that time. However I wasn't using it with cmake so I can't be certain.

If you have a runnable stand-alone bundle (unlike me), you can use this command to see if there are missing frameworks/libraries:
`DYLD_PRINT_LIBRARIES_POST_LAUNCH=1 DYLD_PRINT_LIBRARIES=1 DYLD_PRINT_RPATHS=1 ./nomacs.app/Contents/MacOS/nomacs 2>&1| grep /usr/local`

In the meantime I will look into updating my build environment. I want to stick with OS X 11 since it would theoretically have the broadest compatibility.

```
./nomacs.app/Contents/MacOS/nomacs 
dyld: Library not loaded: @rpath/QtDBus.framework/Versions/A/QtDBus
  Referenced from: /Users/scrubs/nfs/sw/nomacs-testing/hmueller01-nomacs/build-osx6/nomacs.app/Contents/Frameworks/QtGui.framework/Versions/A/QtGui
  Reason: image not found
Abort trap: 6
```

```
[  0%] Built target nomacsCore-3.19.0_autogen_timestamp_deps
[  1%] Built target nomacsCore-3.19.0_autogen
[  2%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkDialog.cpp.o
[  3%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkVersion.cpp.o
[  4%] Linking CXX shared library libnomacsCore-3.19.0.dylib
[ 60%] Built target nomacsCore-3.19.0
[ 60%] Built target nomacs_autogen_timestamp_deps
[ 61%] Built target nomacs_autogen
[ 62%] Building CXX object CMakeFiles/nomacs.dir/src/main.cpp.o
[ 64%] Linking CXX executable nomacs.app/Contents/MacOS/nomacs
[ 98%] Built target nomacs
[100%] Execute /usr/local/bin/macdeployqt to create macOS bundle
ERROR: no file at "/usr/lib/libopencv_imgproc.2.4.dylib"
ERROR: no file at "/usr/lib/libopencv_core.2.4.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/platforms/libqcocoa.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/styles/libqmacstyle.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqcertonlybackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqopensslbackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqsecuretransportbackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/networkinformation/libqglib.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/networkinformation/libqscnetworkreachability.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqgif.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqicns.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqico.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqjp2.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqjpeg.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqmacheif.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqmng.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqpdf.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqsvg.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqtga.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqtiff.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqwbmp.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqwebp.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/iconengines/libqsvgicon.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/platforminputcontexts/libqtvirtualkeyboardplugin.dylib"
WARNING:
WARNING: "nomacs.app/Contents/Resources/qt.conf" already exists, will not overwrite.
WARNING: To make sure the plugins are loaded from the correct location,
WARNING: please make sure qt.conf contains the following lines:
WARNING: [Paths]
WARNING:   Plugins = PlugIns
ERROR: Bundle creation error: "hdiutil: create failed - File exists\n"
[100%] Built target bundle
```

**Comment by hmueller01** [comment]
Hm, ok. The old MacBuildTarget.cmake didn't create a working app on my device. I have macOS Sonoma 14.5, XCode 15.3 and Homebrew 4.3.12.

But something went wong on your machine with these libs
```
ERROR: no file at "/usr/lib/libopencv_imgproc.2.4.dylib"
ERROR: no file at "/usr/lib/libopencv_core.2.4.dylib"
```
I doubt that they are located there. Mine are located under `/usr/local/lib` and will be copied to `nomacs.app/Contents/Frameworks` by macdeployqt.
But as written I have some trouble with qt5 not locating the include, which worked on my old PR. I am no expert in CMake, so it is hard for me to understand what happens where (e.g. creating include paths) ...

Giving some more insights on my build process:
```
> make bundle
[  0%] Built target nomacsCore-3.19.0_autogen_timestamp_deps
[  0%] Built target nomacsCore-3.19.0_autogen
[  0%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkGui/DkDialog.cpp.o
[  1%] Building CXX object CMakeFiles/nomacsCore-3.19.0.dir/src/DkCore/DkVersion.cpp.o
[  2%] Linking CXX shared library libnomacsCore-3.19.0.dylib
ld: warning: ignoring duplicate libraries: '-lc++'
[ 60%] Built target nomacsCore-3.19.0
[ 60%] Built target nomacs_autogen_timestamp_deps
[ 62%] Built target nomacs_autogen
[ 63%] Building CXX object CMakeFiles/nomacs.dir/src/main.cpp.o
[ 63%] Linking CXX executable nomacs.app/Contents/MacOS/nomacs
ld: warning: ignoring duplicate libraries: '-lc++'
[ 98%] Built target nomacs
[100%] Execute /usr/local/bin/macdeployqt to create macOS bundle
[100%] Built target bundle
```
So no critical outputs.

This `File exists, skip copy:` outputs result if you did run `make bundle` before.
As well as `ERROR: Bundle creation error: "hdiutil: create failed - File exists\n"`
which means that `nomacs.dmg ` exists (and won't be overwritten).

But it looks like there are still some unresolved libs, but I do not understand why ...
 ```
DYLD_PRINT_LIBRARIES_POST_LAUNCH=1 DYLD_PRINT_LIBRARIES=1 DYLD_PRINT_RPATHS=1 ./nomacs.app/Contents/MacOS/nomacs 2>&1| grep /usr/local
dyld[75155]: <634D5C01-3B5E-3BC9-8D1C-47BCBBCC8CA1> /usr/local/Cellar/opencv/4.10.0/lib/libopencv_core.4.10.0.dylib
dyld[75155]: <8DAB0FAA-1CA6-3BC9-9EF3-4C68889B07B8> /usr/local/Cellar/openblas/0.3.27/lib/libopenblasp-r0.3.27.dylib
dyld[75155]: <5D8A55D0-B783-34B3-8B1E-EFF97D0AE534> /usr/local/Cellar/tbb/2021.13.0/lib/libtbb.12.13.dylib
dyld[75155]: <7A20A672-BEC7-3F96-BDCE-7CAF46B05270> /usr/local/Cellar/gcc/14.1.0_2/lib/gcc/current/libgfortran.5.dylib
dyld[75155]: <8652735D-1220-3F1C-B1FA-2F2045297E71> /usr/local/Cellar/gcc/14.1.0_2/lib/gcc/current/libgomp.1.dylib
dyld[75155]: <9003EA36-ED2C-34EF-A914-FE93854380FC> /usr/local/Cellar/gcc/14.1.0_2/lib/gcc/current/libquadmath.0.dylib
dyld[75155]: <96441B6C-CDAD-3735-8344-338C19F1CD82> /usr/local/Cellar/gcc/14.1.0_2/lib/gcc/current/libgcc_s.1.1.dylib
dyld[75155]: <645B37DE-FED9-34A0-9587-C79CBDB3EE81> /usr/local/Cellar/tbb/2021.13.0/lib/libtbbmalloc.2.13.dylib
```

otool outputs:
```
otool -L nomacs.app/Contents/MacOS/nomacs 
nomacs.app/Contents/MacOS/nomacs:
	@rpath/libnomacsCore-3.19.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	@executable_path/../Frameworks/libexiv2.28.dylib (compatibility version 28.0.0, current version 0.28.3)
	@executable_path/../Frameworks/libraw.23.dylib (compatibility version 24.0.0, current version 24.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1700.255.0)
	@executable_path/../Frameworks/liblcms2.2.dylib (compatibility version 3.0.0, current version 3.16.0)
	@executable_path/../Frameworks/libopencv_imgproc.410.dylib (compatibility version 410.0.0, current version 4.10.0)
	@executable_path/../Frameworks/libtiff.6.dylib (compatibility version 7.0.0, current version 7.2.0)
	@executable_path/../Frameworks/libopencv_core.410.dylib (compatibility version 410.0.0, current version 4.10.0)
	@executable_path/../Frameworks/QtNetwork.framework/Versions/A/QtNetwork (compatibility version 6.0.0, current version 6.7.0)
	@executable_path/../Frameworks/QtPrintSupport.framework/Versions/A/QtPrintSupport (compatibility version 6.0.0, current version 6.7.0)
	@executable_path/../Frameworks/QtWidgets.framework/Versions/A/QtWidgets (compatibility version 6.0.0, current version 6.7.0)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 64.0.0)
	/usr/lib/libcups.2.dylib (compatibility version 2.0.0, current version 2.14.0)
	@executable_path/../Frameworks/QtConcurrent.framework/Versions/A/QtConcurrent (compatibility version 6.0.0, current version 6.7.0)
	@executable_path/../Frameworks/QtSvg.framework/Versions/A/QtSvg (compatibility version 6.0.0, current version 6.7.0)
	@rpath/QtGui.framework/Versions/A/QtGui (compatibility version 6.0.0, current version 6.7.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2487.50.124)
	/System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 343.14.0)
	@executable_path/../Frameworks/QtCore.framework/Versions/A/QtCore (compatibility version 6.0.0, current version 6.7.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
```
which looks good from my point of view.

File: README.md (hunk 1 above)
Comment (hmueller01) [reply README.md:152]: True. The nomas.app is working with references to the build environment. I'll change this.

**Comment by hmueller01** [comment]
Ok, qt5 build works as well. I had written it in the README already ...
Had to add the
`export CPLUS_INCLUDE_PATH=/usr/local/include`
before running cmake.

**Comment by scrubbbbs** [comment]
> But something went wong on your machine with these libs

Yes, it's my fault! I have opencv 2.4 compiled without homebrew in /usr/local.

> But it looks like there are still some unresolved libs, but I do not understand why ...

- libnomacscore also has dependencies
- anything loaded at runtime (qt plugins, nomacs plugins)

It may be necessary to manually had some libs. linuxdeployqt has argument for this,  maybe the homebrew version oof macdeployqt does as well.

There are a lot of open bugs for macdeployqt, I have a feeling this is why I rolled my own bundler.

**Comment by scrubbbbs** [comment]
I have a new build environment using MacOSX v12, both Qt5 and Qt6 x86 builds (all I can test) produced a working bundle.

Qt5: I needed to set CMAKE_PREFIX_PATH as cmake wasn't finding qt5 cmake file.  `make bundle` ran cleanly but did not produce a stand-alone bundle.

Qt6: Compiles cleanly, but `make bundle` gives a ton of warnings/errors. The resulting bundle is still runnable, but it is missing similar set of libraries to the qt5 version.

**Comment by hmueller01** [comment]
@scrubbbbs Thanks for merging. 
We could add  `-always-overwrite` to `macdeployqt` to avoid the error messages you stated ...

```
[100%] Execute /usr/local/bin/macdeployqt to create macOS bundle
File exists, skip copy: "nomacs.app/Contents/PlugIns/platforms/libqcocoa.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/styles/libqmacstyle.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqcertonlybackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqopensslbackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/tls/libqsecuretransportbackend.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/networkinformation/libqglib.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/networkinformation/libqscnetworkreachability.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqgif.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqicns.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqico.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqjp2.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqjpeg.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqmacheif.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqmng.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqpdf.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqsvg.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqtga.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqtiff.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqwbmp.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/imageformats/libqwebp.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/iconengines/libqsvgicon.dylib"
File exists, skip copy: "nomacs.app/Contents/PlugIns/platforminputcontexts/libqtvirtualkeyboardplugin.dylib"
WARNING:
WARNING: "nomacs.app/Contents/Resources/qt.conf" already exists, will not overwrite.
WARNING: To make sure the plugins are loaded from the correct location,
WARNING: please make sure qt.conf contains the following lines:
WARNING: [Paths]
WARNING:   Plugins = PlugIns
ERROR: Bundle creation error: "hdiutil: create failed - Datei existiert\n"
[100%] Built target bundle
```

**Comment by scrubbbbs** [comment]
> We could add `-always-overwrite` to `macdeployqt` to avoid the error messages you stated ...

Thanks, and I would add it except it seems like `macdeployqt` isn't going to work (for standalone build) unless I'm missing something...

To achieve parity with other platforms we will need:
- github workflow for arm and x86 versions to validate PRs
- simple command or automation to push new releases (both stable and  unstable) . I'm not sure how this works currently with windows build

It's the second part where we have a problem due to no standalone build. We seem to have enough to validate PRs now.

**Pull request merged**